### ISSUE TITLE
Allow process monitoring

### ISSUE THREAD

**Issue by ranjib** [influxdb/telegraf#215]
Gathering stats about a particular process' memory, cpu consumption is easy, less invasive and generic way of  monitoring any system. It will be nice to have a plugin that allows users to specify a process via pidfile or executable name and gathers it memory, cpu consumption data.
This is very similar to Nagios' [check_procs](https://www.monitoring-plugins.org/doc/man/check_procs.html)

**Comment by sparrc** [comment]
This is definitely something that we need, thanks for making an issue to track it

**Comment by ranjib** [comment]
@sparrc i can take a stab at it. It will be nice to capture all the details (if you have anything specific in mind). I am thinking of how to specify the expected process, i.e. via pid file or via executable path or via dbus calls.

**Comment by sparrc** [comment]
@ranjib That would be great and much appreciated!

As for the filtering, I would say that we probably just want all (or at least most) of the same filters that the nagios check_procs plugin has:

```
Filters:
 -s, --state=STATUSFLAGS
   Only scan for processes that have, in the output of `ps`, one or
   more of the status flags you specify (for example R, Z, S, RS,
   RSZDT, plus others based on the output of your 'ps' command).
 -p, --ppid=PPID
   Only scan for children of the parent process ID indicated.
 -z, --vsz=VSZ
   Only scan for processes with VSZ higher than indicated.
 -r, --rss=RSS
   Only scan for processes with RSS higher than indicated.
 -P, --pcpu=PCPU
   Only scan for processes with PCPU higher than indicated.
 -u, --user=USER
   Only scan for processes with user name or ID indicated.
 -a, --argument-array=STRING
   Only scan for processes with args that contain STRING.
 --ereg-argument-array=STRING
   Only scan for processes with args that contain the regex STRING.
 -C, --command=COMMAND
   Only scan for exact matches of COMMAND (without path).
 -k, --no-kthreads
   Only scan for non kernel threads (works on Linux only).
```

**Comment by gotyaoi** [comment]
I would definitely appreciate a plugin like this. Another source of inspiration might be the processes plugin from collectd, which by default records the number of processes grouped by state, and then will provide more information about processes specified by name or regex. https://collectd.org/wiki/index.php/Plugin:Processes

**Comment by aderumier** [comment]
Hi,on elastic camp, they have a new tool to get process metrics (cpu/memory usage)

https://www.elastic.co/guide/en/beats/topbeat/current/topbeat-getting-started.html

**Comment by ranjib** [comment]
@sparrc i have few question regarding the measurement we want to get out of this plugin. If we filter processes or rss/vms values, what are the metrics/measurements we want to publish? Total number of processes? Do we want total value of RSS/VMS by all processes filtered by the criteria ?
Following is a feature request from https://github.com/influxdb/telegraf/issues/235
- total number of system processes/threads (by state - runnable,sleeping,zombie)

**Comment by sparrc** [comment]
closing this one as fixed by the `procstat` plugin, will use #235 to track the total processes count
